pyroomacoustics/bss/fastmnmf.py

```diff
@@ -0,0 +1,168 @@
+    initialize_ilrma=False
+):
+    """
+    Implementation of FastMNMF algorithm presented in
+
+    K. Sekiguchi, A. A. Nugraha, Y. Bando, K. Yoshii, *Fast Multichannel Source 
+    Separation Based on Jointly Diagonalizable Spatial Covariance Matrices*, EUSIPCO, 2019. [`arXiv <https://arxiv.org/abs/1903.03237>`_]
+
+    The code of FastMNMF with GPU support and FastMNMF-DP which integrates DNN-based source model
+    into FastMNMF is available on https://github.com/sekiguchi92/SpeechEnhancement 
+
+    Parameters
+    ----------
+    X: ndarray (nframes, nfrequencies, nchannels)
+        STFT representation of the observed signal 
+    n_src: int, optional
+        The number of sound sources (if n_src = None, n_src is set to the number of microphone)
+    n_iter: int, optional
+        The number of iterations
+    W0: ndarray (nfrequencies, nchannels, nchannels), optional
+        Initial value for diagonalizer Q
+        Demixing matrix can be used as the initial value
+    n_components: int, optional
+        Number of components in the non-negative spectrum
+    callback: func, optional
+        A callback function called every 10 iterations, allows to monitor convergence
+    mic_index: int, optional
+        The index of microphone of which you want to get the source image
+    interval_update_Q: int, optional
+        The interval of updating Q
+    interval_normalize: int, optional
+        The interval of parameter normalization
+    initialize_ilrma: boolean, optional
+        Initialize diagonalizer Q by using ILRMA
+
+    Returns
+    -------
+    separated spectrogram: numpy.ndarray
+        An (nframes, nfrequencies, nsources) array. 
+    """
+
+    # initialize parameter
+    X_FTM = X.transpose(1, 0, 2)
+    n_freq, n_frames, n_chan = X_FTM.shape
+    XX_FTMM = np.matmul(X_FTM[:, :, :, None], X_FTM[:, :, None, :].conj())
+    if n_src is None:
+        n_src = X_FTM.shape[2] # determined case (the number of source = the number of microphone)
+
+    covarianceDiag_NFM = np.random.rand(n_src, n_freq, n_chan)
+    power_observation_FT = (np.abs(X_FTM) ** 2).mean(axis=2) # F T
+    if initialize_ilrma: # initialize by using ILRMA
+        from pyroomacoustics.bss.ilrma import ilrma
+        _, W = ilrma(X, n_iter=10, n_components=2, proj_back=False, return_filters=True)
+        diagonalizer_FMM = W
+        covarianceDiag_NFM[0] = 1e-4
+        covarianceDiag_NFM[0, :, 0] = 1
+    elif W0 != None: # initialize by W0
+        diagonalizer_FMM = W0
+    else: # initialize by using observed signals
+        covarianceMatrix_FMM = XX_FTMM.sum(axis=1) / power_observation_FT.sum(axis=1)[:, None, None] # F M M
+        covarianceMatrix_FMM = covarianceMatrix_FMM / np.trace(covarianceMatrix_FMM, axis1=1 ,axis2=2)[:, None, None]
+        eig_val, eig_vec = np.linalg.eig(covarianceMatrix_FMM)
+        diagonalizer_FMM = eig_vec.transpose(0, 2, 1).conj()
+        covarianceDiag_NFM[1] = eig_val.real
+    mu_NF = np.zeros([n_chan, n_freq])
```

Comment:
Thank you for the great work. I ran the test and seems there are no problems with this commit. I just noticed this line is not really doing anything, so we could remove it before merging.

Reply:
What you are saying is correct. I'm sorry.

Reply:
Hi @sekiguchi92, no need to apologize! The code review is meant to help get the code merged into master, not as a critique. Your contribution is much appreciated! 👍